lib/cn/kblock_builder.c

```diff
@@ -43,7 +43,13 @@
 
 extern struct tbkt sp3_tbkt;
 
-#define HSP_HASH_MAX_KEYS (16 * 1024)
+/* The max number of keys in a struct hash_set_part is sized such that the allocation
+ * size of a hash_set_part is one page less than VLB_ALLOCSZ_MAX so that vlb_free()
+ * can cache it.  This eliminates many trips into the kernel that would otherwise
+ * have to modify the address map.
+ */
+#define HSP_HASH_MAX_KEYS \
+    ((VLB_ALLOCSZ_MAX - 4096 - sizeof(struct hash_set_part)) / sizeof(((struct hash_set_part *)0))->hashvec[0])
```

Comment:
Is this 4096 supposed to be a page size or 4M? Also we have an NELEM macro in hse_util/base.h, which looks like the last part is?

Reply:
No, it's just the size of the red-zone I gave to vlb_free() for data it stores in that page.  So if the user overruns that barrier then  I cannot trust the data and I just free (munmap) the buffer.  But 4096 is synonymous with page size that they're often conflated, like what I did in a comment somewhere...

4M is the alloction size (mmap).  NELEM won't work here because the array is a flexible array.  I don't want the size of the array, I'm computing it!  To do that I need the size of one element....

Reply:
Would be nice to have a constant for the red-zone and ideally a static_assert() somewhere so others can recognize how the values are related.